util/src/main/scala/org/ensime/util/Debouncer.scala

```diff
@@ -0,0 +1,118 @@
+// Copyright: 2010 - 2017 https://github.com/ensime/ensime-server/graphs
+// License: http://www.gnu.org/licenses/gpl-3.0.en.html
+package org.ensime.util
+
+import akka.actor.Scheduler
+import java.time.Clock
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicInteger }
+import org.slf4j.LoggerFactory
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+/**
+ * Generic debouncer implementation
+ *
+ * @param name The name of this debouncer
+ * @param scheduler The scheduler implementation to use; Can accept an akka scheduler a la magnet pattern
+ * @param delay The time after which to invoke action; each time [[Debouncer.call]] is invoked, this delay is reset
+ * @param maxDelay The maximum time to delay the invocation of action()
+ * @param action The action to invoke
+ * @param debounceFirst If false, then the first invocation of call() calls action exactly once. Subsequent calls are debounced.
+ * @param Clock The source from which we get the current time. This input should use the same source. Specified for testing purposes
+ */
+class Debouncer(
+    name: String,
+    scheduler: Debouncer.SchedulerLike,
+    delay: FiniteDuration,
+    maxDelay: FiniteDuration,
+    action: () => Unit,
+    debounceFirst: Boolean = true,
+    clock: Clock = Clock.systemUTC
+) {
+  require(delay <= maxDelay, "delay should be <= maxDelay")
+  private val logger = LoggerFactory.getLogger(getClass)
+  private[this] val pending = new AtomicBoolean(false)
+  private[this] val calls = new AtomicInteger(0)
+  private[this] val debouncingActive = new AtomicBoolean(debounceFirst)
+  private val delayMs = delay.toMillis
+  private val maxDelayMs = maxDelay.toMillis
+  @volatile private[this] var lastCallAttempt: Long = clock.millis()
+  @volatile private[this] var lastInvocation: Long = clock.millis()
+
+  /**
+   * Register that the provided action should be called according to the debounce logic
+   */
+  def call(): Unit = if (debouncingActive.compareAndSet(false, true)) {
+    // debounceFirst was false
```

Comment:
another way to do this could be to reverse the meaning of the initial boolean flag... then the logic stays the same without double negation. `runFirstMessage` with `debouncingActive = !runFirstMessage`